cmake/CeleritasUtils.cmake

```diff
@@ -43,15 +43,180 @@ macro(celeritas_find_package_config _package)
+    INTERFACE_LINK_LIBRARIES )
+  set_target_properties(${target}_static PROPERTIES
+    LINK_LIBRARIES "${target_LINK_LIBRARIES}"
+    INTERFACE_LINK_LIBRARIES "${target_INTERFACE_LINK_LIBRARIES}"
+  )
+
+  set_target_properties(${target}_static PROPERTIES
+    POSITION_INDEPENDENT_CODE ON
+  )
+endif()
+
 endfunction()
 
+
+define_property(TARGET PROPERTY CELERITAS_CUDA_LIBRARY_TYPE
+  BRIEF_DOCS "Indicate the type of cuda library (STATIC and SHARED for nvlink usage, FINAL for linking into not cuda library/executable"
+  FULL_DOCS "Indicate the type of cuda library (STATIC and SHARED for nvlink usage, FINAL for linking into not cuda library/executable"
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_FINAL_LIBRARY
+  BRIEF_DOCS "Name of the final library corresponding to this cuda library"
+  FULL_DOCS "Name of the final library corresponding to this cuda library"
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_STATIC_LIBRARY
+  BRIEF_DOCS "Name of the static library corresponding to this cuda library"
+  FULL_DOCS "Name of the static library corresponding to this cuda library"
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_MIDDLE_LIBRARY
+  BRIEF_DOCS "Name of the shared (without nvlink step) library corresponding to this cuda library"
+  FULL_DOCS "Name of the shared (without nvlink step) library corresponding to this cuda library"
+)
+
+function(celeritas_add_library target)
+  if(NOT BUILD_SHARED_LIBS OR NOT CELERITAS_USE_CUDA)
+    list(SUBLIST ARGV 1 -1 NEWARGV)
+    add_library(${target} ${NEWARGV})
+    add_library(${target}_final ALIAS ${target})
+    add_library(${target}_cuda ALIAS ${target})
+    add_library(${target}_static ALIAS ${target})
+    return()
+  endif()
+
+  list(SUBLIST ARGV 1 -1 NEWARGV)
+
+  add_library(${target}_objects OBJECT ${NEWARGV})
+  add_library(${target}_static STATIC $<TARGET_OBJECTS:${target}_objects>)
+  add_library(${target}_cuda SHARED $<TARGET_OBJECTS:${target}_objects>)
+  # We need to use a dummy file as a library (per cmake) needs to contains
+  # at least one source file.  The real content of the library will be
+  # the cmake_device_link.o resulting from the execution of `nvcc -dlink`
+  # Also non-cuda related test, for example `gtest_detail_Macros`,
+  # will need to be linked again libceleritas_final while a library
+  # that the detends on and that uses Celeritas::Core (for example
+  # libCeleritasTest.so) will need to be linked against `libceleritas_cuda`.
+  # If both the `_cuda` and `_final` contains the `.o` files we would
+  # then have duplicated symbols (Here the symptoms will a crash
+  # during the cuda library initialization rather than a link error).
+  add_library(${target}_final SHARED ../src/base/dummy.cu)
+
+  set_target_properties(${target}_objects PROPERTIES
+    POSITION_INDEPENDENT_CODE ON
+    CUDA_SEPARABLE_COMPILATION ON
+    CUDA_RUNTIME_LIBRARY Shared
+  )
+
+  set_target_properties(${target}_cuda PROPERTIES
```

Comment:
It seems to me that this library doesn't necessarily have CUDA code in it: it's just a shared library that includes the same objects as the static lib. The `_cuda` nomenclature is confusing e.g. for ROOT linkage case, where nothing in the ROOT interfaces are interacting with CUDA; it's just that ROOT needs to link against the shared libraries.

Of course, `_shared` will also be a misnomer if CUDA+VecGeom are disabled... 🤔

Reply:
If cuda is disable, the current code does not produce any of `_[cuda/init/final]`.  As noted on line 159ish, this could also be done automatically if cuda is enabled but the library does not contain cuda code (but the current "heureustic" to guess that is brittle (see you comment on .cuda.cc).  I suppose we could also try to disable the mechanism (and thus separatable compilation) if VecGeom is not enabled.

Reply:
It'll still create an alias library `celeritas_cuda`, right?  I came back to this point because I noticed that `celeritas::ROOT` changed from linking against `celeritas` to `celeritas_cuda`, which is odd.

Reply:
> against celeritas to celeritas_cuda, which is odd.

This is odd sounding per se but is still accurate (in this case `celeritas::ROOT` need to link aginst the code Celeritas library which contains both the cxx coda and the cuda code (and technically could be calling/using both)).

Nonetheless, I have been thinking of renaming that library to either `[celeritas]_middle` or `[celeritas]_intermediary`.